File: Problems/199.binary-tree-right-side-view.py
# title: Binary Tree Right Side View
# timestamp: 2024-12-21 10:05:31
# problemUrl: https://leetcode.com/problems/binary-tree-right-side-view/
# memory: 18 MB
# runtime: 0 ms

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        lvl, newLvl, res = [root], [], []

        while lvl:
            res.append(lvl[-1].val)
            for node in lvl:
                if node.left:
                    newLvl.append(node.left)
                if node.right:
                    newLvl.append(node.right)
            lvl, newLvl = newLvl, []
        
        return res

File: Problems/199.binary-tree-right-side-view.go
// title: Binary Tree Right Side View
// timestamp: 2024-04-18 21:37:26
// problemUrl: https://leetcode.com/problems/binary-tree-right-side-view/
// memory: 2.6 MB
// runtime: 0 ms

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
    if root == nil {
        return nil
    }

    var level, newLevel []*TreeNode
    level = []*TreeNode{root}

    var view []int
    for len(level) != 0 {
        view = append(view, level[len(level) - 1].Val)
        
        for _, node := range level {
            if node.Left != nil {
                newLevel = append(newLevel, node.Left)
            }
            if node.Right != nil {
                newLevel = append(newLevel, node.Right)
            }
        }
        level = newLevel
        newLevel = nil
    }

    return view
}